sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -608,12 +608,16 @@ case class FileSourceScanExec(
     new FileScanRDD(fsRelation.sparkSession, readFile, partitions)
   }
 
+  private def filterUnnecessaryPredicates(predicates: Seq[Expression]): Seq[Expression] = {
+    predicates.filterNot(_ == Literal.TrueLiteral)
+  }
+
   override def doCanonicalize(): FileSourceScanExec = {
     FileSourceScanExec(
       relation,
       output.map(QueryPlan.normalizeExpressions(_, output)),
       requiredSchema,
-      QueryPlan.normalizePredicates(partitionFilters, output),
+      QueryPlan.normalizePredicates(filterUnnecessaryPredicates(partitionFilters), output),
```

Comment:
Why does `TrueLiteral` affect canonicalize? `TrueLiteral == TrueLiteral` returns true, right?

Reply:
Ah I got it. Maybe one side has DPP and one side not.

 Can we filter out `DynamicPruningExpression(Literal.TrueLiteral)`? Then we don't need to change https://github.com/apache/spark/pull/29318/files#diff-6c00f36fadec282fce72ad39cfea51caL81 and make this PR less risky.

Reply:
done.